### ISSUE TITLE
Tarantool crashes with error "not enough memory"

### ISSUE THREAD

**Issue by derElektrobesen** [tarantool/tarantool#4374]
**Tarantool version:**
1.10.3-95-ga7d003a

**OS version:**
Centos 7

**Description:**

We have following tarantools configuration:
* 4 phisical servers, 2 for masters and 2 for slaves
* 4 tarantool shards per server
* 30G arena for each tarantool
* 256G of phisical memory per server
* Vinyl not used

Tarantool just crashes with the following lines in log file:
```
2019-07-21 19:52:03.899 [9278] main/917292/main txn.c:314 W> too long WAL write: 1 rows at LSN 18783407441: 0.558 sec
2019-07-21 19:52:13.849 [9278] main/113/fifottl txn.c:314 W> too long WAL write: 1 rows at LSN 18783429547: 0.512 sec
2019-07-21 19:52:18.870 [9278] main/113/fifottl txn.c:314 W> too long WAL write: 1 rows at LSN 18783440812: 0.531 sec
2019-07-21 19:52:51.690 [9278] relay/10.161.198.4:44254/101/main I> done `xlogs/00000000018782514162.xlog'
2019-07-21 19:52:51.690 [9278] relay/10.161.198.4:44254/101/main I> recover from `xlogs/00000000018783514162.xlog'
2019-07-21 19:52:58.911 [9278] main/918195/main txn.c:314 W> too long WAL write: 1 rows at LSN 18783528851: 0.568 sec
2019-07-21 19:53:48.987 [9278] main/917986/main txn.c:314 W> too long WAL write: 1 rows at LSN 18783638951: 0.643 sec
2019-07-21 19:54:08.852 [9278] main/918109/main txn.c:314 W> too long WAL write: 1 rows at LSN 18783682827: 0.505 sec
2019-07-21 19:55:12.333 [9278] main/918328/main C> not enough memory
```

There are very many warnings with `too long WAL write` in the log. But atop doesn't show any problems with discs.

It seems there is a leak in tarantool.

Following picture shows tarantool memory usage in the moment of crash:
![Снимок экрана 2019-07-22 в 12 30 19](https://user-images.githubusercontent.com/4015637/61621991-99eed880-ac7c-11e9-8d39-ceb47259dd2f.png)

CPU usage in the same moment:
![screenshot](https://user-images.githubusercontent.com/4015637/61622092-cd316780-ac7c-11e9-94ea-7149e3a20813.png)

This shard crashed just after 5 days it started:
![screenshot](https://user-images.githubusercontent.com/4015637/61622763-17671880-ac7e-11e9-9bf3-cdc0d60767e5.png)

We disabled LuaJIT on one tarantool shard (using `jit.off()`) to test the leak. But it seems there is the same problem on this shard too:
![Снимок экрана 2019-07-22 в 12 35 54](https://user-images.githubusercontent.com/4015637/61622373-5183ea80-ac7d-11e9-8512-71c3812b8347.png)
```
jit.status()
---
- false
- SSE2
- SSE3
- SSE4.1
- BMI2
- fold
- cse
- dce
- fwd
- dse
- narrow
- loop
- abc
- sink
- fuse
...
```

We tried to use pprof to find a problem, it shown the following.

On common shard:
```
+   98.82%     0.00%  tarantool        [unknown]                              [k] 0x40d67ab041c863b8
+   98.78%     0.00%  tarantool        tarantool                              [.] lj_BC_FUNCC
+   98.38%     0.00%  tarantool        tarantool                              [.] lj_cf_ffi_gc
+   98.38%     0.00%  tarantool        tarantool                              [.] lj_cdata_setfin
+   98.37%     0.00%  tarantool        tarantool                              [.] lj_tab_newkey
+   98.34%    96.23%  tarantool        tarantool                              [.] rehashtab
+    1.96%     0.00%  tarantool        [kernel.kallsyms]                      [k] system_call_fastpath
+    1.44%     0.01%  tarantool        tarantool                              [.] lj_alloc_malloc
+    1.40%     0.00%  tarantool        tarantool                              [.] lj_tab_resize
+    1.38%     0.01%  tarantool        tarantool                              [.] lj_mem_realloc
+    1.33%     0.00%  tarantool        libc-2.17.so;5d14bda2 (deleted)        [.] 0xffff804681c430f7
+    1.17%     0.00%  tarantool        [kernel.kallsyms]                      [k] sys_munmap
+    1.16%     0.01%  tarantool        [kernel.kallsyms]                      [k] vm_munmap
+    1.14%     0.04%  tarantool        [kernel.kallsyms]                      [k] do_munmap
+    0.94%     0.00%  tarantool        libc-2.17.so;5d14bda2 (deleted)        [.] 0xffff804681c4306a
+    0.79%     0.01%  tarantool        [kernel.kallsyms]                      [k] sys_mmap_pgoff
+    0.78%     0.00%  tarantool        [kernel.kallsyms]                      [k] sys_mmap
+    0.77%     0.01%  tarantool        [kernel.kallsyms]                      [k] vm_mmap_pgoff
+    0.74%     0.01%  tarantool        [kernel.kallsyms]                      [k] do_mmap
+    0.67%     0.01%  tarantool        [kernel.kallsyms]                      [k] unmap_region
     0.50%     0.02%  tarantool        [kernel.kallsyms]                      [k] mmap_region
```

On shard with disabled LuaJIT:
```
+   92.71%     0.00%  tarantool        tarantool           [.] lj_gc_step
+   92.71%     0.00%  tarantool        tarantool           [.] lj_gc_step_jit
+   92.71%     0.00%  tarantool        perf-21043.map      [.] 0x000000002a5f6940
+   92.71%     0.00%  tarantool        [unknown]           [.] 0x003300000017001a
+   92.71%    92.71%  tarantool        tarantool           [.] propagatemark
+    6.21%     0.00%  tarantool        tarantool           [.] box_iterator_next
+    6.21%     0.00%  tarantool        tarantool           [.] tree_iterator_next_equal
+    6.21%     0.00%  tarantool        perf-21043.map      [.] 0x000000002a5f6aeb
+    6.21%     6.21%  tarantool        tarantool           [.] tuple_field_raw
+    1.07%     0.00%  tarantool        [unknown]           [.] 0x4138df3841a3f3b8
+    1.07%     0.00%  tarantool        tarantool           [.] lj_vmeta_tgetv
+    1.07%     0.00%  tarantool        tarantool           [.] lj_meta_tget
+    1.07%     0.00%  tarantool        tarantool           [.] lj_meta_lookup
+    1.07%     1.07%  tarantool        tarantool           [.] lj_tab_getstr
     0.00%     0.00%  tarantool        tarantool           [.] lj_BC_FUNCC
```

**Comment by derElektrobesen** [comment]
Tarantool in the moment of crash shows following stack-trace:
```
Jun 22 19:20:14 klumba1 tarantoolctl: #0  0x4f7829 in print_backtrace+9
Jun 22 19:20:14 klumba1 tarantoolctl: #1  0x4dc429 in tarantool_panic_handler+29
Jun 22 19:20:14 klumba1 tarantoolctl: #2  0x5147b3 in lj_err_throw+73
Jun 22 19:20:14 klumba1 tarantoolctl: #3  0x5149ab in lj_err_mem+3b
Jun 22 19:20:14 klumba1 tarantoolctl: #4  0x514408 in lj_mem_realloc+48
Jun 22 19:20:14 klumba1 tarantoolctl: #5  0x51a42d in stack_init+1d
Jun 22 19:20:14 klumba1 tarantoolctl: #6  0x51a9eb in lj_state_new+5b
Jun 22 19:20:14 klumba1 tarantoolctl: #7  0x52156f in lua_newthread+1f
Jun 22 19:20:14 klumba1 tarantoolctl: #8  0x4ccf9f in box_lua_call+1f
Jun 22 19:20:14 klumba1 tarantoolctl: #9  0x4cb1af in box_process_call+12f
Jun 22 19:20:14 klumba1 tarantoolctl: #10 0x4122b7 in _ZL15tx_process_callP4cmsg+d7
Jun 22 19:20:14 klumba1 tarantoolctl: #11 0x4f90f4 in cmsg_deliver+14
Jun 22 19:20:14 klumba1 tarantoolctl: #12 0x4f9cbf in fiber_pool_f+ff
Jun 22 19:20:14 klumba1 tarantoolctl: #13 0x40f07c in _ZL16fiber_cxx_invokePFiP13__va_list_tagES0_+c
Jun 22 19:20:14 klumba1 tarantoolctl: #14 0x4f4510 in fiber_loop+30
Jun 22 19:20:14 klumba1 tarantoolctl: #15 0x62dd7f in coro_init+3f
```

**Comment by Totktonada** [comment]
Looks similar to #562.

**Comment by igormunkin** [comment]
Looks to be a #4556 sibling (closed via 6f8ce4b).
